### ISSUE TITLE
[question] Can we use different AMI images for different instance arch in mixed IGs?

### ISSUE THREAD

**Issue by shapirus** [kubernetes/kops#16100]
There's a scalar field to specify the image AMI for an instance group: `.spec.image`

However, it is sometimes desirable to run instances of different architectures (arm64 + amd64) in the same instance group, using, for example, the spot mixedInstancesPolicy configuration.

Obviously, this will fail, since all the instance types must match the specified AMI architecture.

Part of the documentation at https://kops.sigs.k8s.io/instance_groups/#spotinstancepools suggests that there might be something of interest:

> The Spot pools are determined from the different instance types in the Overrides array of LaunchTemplate. Default if not set is 2.

However I found no further mention of how to configure this Overrides array, which apparently can be used to set AMIs for each specific instance type (as described in https://aws.amazon.com/blogs/compute/supporting-aws-graviton2-and-x86-instance-types-in-the-same-auto-scaling-group/).

Is there any workaround?

**Comment by jclegras** [comment]
Hello, is there anything new here please ?

**Comment by shapirus** [comment]
/remove-lifecycle rotten

maybe one day it gets some attention, so let's keep it open.

**Comment by shapirus** [comment]
/remove-lifecycle stale

maybe eventually

**Comment by shapirus** [comment]
/remove-lifecycle stale

this is actually somewhat important as mixed architecture deployments are widely used given the abundance of cross-architecture software builds and cloud instances.

so we should rather keep it on the radar.

**Comment by rifelpet** [comment]
Is maintaining separate IGs per architecture insufficient? As @hakman mentioned, it would be a significant effort to support - and if the workaround of using separate IGs is trivial then I'm not sure if we'll end up prioritizing this.

**Comment by shapirus** [comment]
> Is maintaining separate IGs per architecture insufficient? As [@hakman](https://github.com/hakman) mentioned, it would be a significant effort to support - and if the workaround of using separate IGs is trivial then I'm not sure if we'll end up prioritizing this.

Yes, the workaround with separate IGs would be obvious and trivial, except that it's hardly a workaround: it defeats the purpose and doesn't allow to take advantage of, for example, AWS mixed spot instance IGs with multiple instance types. Splitting the IG into two makes it less flexible in terms of costs and spot capacity & interruption handling.

On the other hand, is it even supported on the AWS side? It would require creating individual launch templates for each architecture or two different AMIs in the same LT for different architectures, and I don't really know if AWS supports it. Not sure of other hosting providers either.

I'm not speaking of prioritizing this issue of course, just of keeping it afloat for a later time when the stars align properly and implementing it will be feasible without going into much trouble.

**Comment by hakman** [comment]
> On the other hand, is it even supported on the AWS side? It would require creating individual launch templates for each architecture or two different AMIs in the same LT for different architectures, and I don't really know if AWS supports it. Not sure of other hosting providers either.

It is supported on AWS side, but not sure on other providers. It would be more of a hack.
https://docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-mixed-instances-groups-launch-template-overrides.html

> I'm not speaking of prioritising this issue of course, just of keeping it afloat for a later time when the stars align properly and implementing it will be feasible without going into much trouble.

This may be in the cards once we get to fix Karpenter support later this year.
